File: backend/internal/dto/invitation.go
package dto

import "time"

type ListInvitationsReq struct {
	CreatorID      *int64     `json:"creator_id"`
	MaxUsedCnt     *int64     `json:"max_used_cnt"`
	ExpiresAtStart *string    `json:"expires_at_start"`
	ExpiresAtEnd   *string    `json:"expires_at_end"`
	CreatedAtStart *string    `json:"created_at_start"`
	CreatedAtEnd   *string    `json:"created_at_end"`
	Disabled       *bool      `json:"disabled"`
	OnlyMine       *bool      `json:"only_mine"`
	Keyword        *string    `json:"keyword"`
	SortArgs       SortArgs   `json:"sort_args"`
	Pagination     Pagination `json:"pagination"`
}

type CreateInvitationRequest struct {
	CreatorExternalID string     `json:"creator_external_id"`
	MaxUsedCnt        *int64     `json:"max_used_cnt"`
	ExpiresAt         *time.Time `json:"expires_at"`
	Note              *string    `json:"note"`
}

type UpdateInvitationRequest struct {
	Code       string     `json:"code"`
	MaxUsedCnt *int64     `json:"max_used_cnt"`
	ExpiresAt  *time.Time `json:"expires_at"`
	Disabled   *bool      `json:"disabled"`
	Note       *string    `json:"note"`
}

type DeleteInvitationRequest struct {
	Code string `json:"code"`
}

type InvitationResponse struct {
	ID                  int64      `json:"id"`
	Code                string     `json:"code"`
	CreatedByExternalID string     `json:"created_by_external_id"`
	CreatedByName       string     `json:"created_by_name"`
	UsedCnt             int64      `json:"used_cnt"`
	MaxUsedCnt          *int64     `json:"max_used_cnt"`
	ExpiresAt           *time.Time `json:"expires_at"`
	CreatedAt           time.Time  `json:"created_at"`
	Disabled            bool       `json:"disabled"`
	Note                *string    `json:"note"`
}

type ListInvitationRecordsRequest struct {
	Code        *string    `json:"code"`
	Status      *string    `json:"status"`
	UsedAtStart *string    `json:"used_at_start"`
	UsedAtEnd   *string    `json:"used_at_end"`
	CreatorID   *int64     `json:"creator_id"`
	OnlyMine    *bool      `json:"only_mine"`
	Keyword     *string    `json:"keyword"`
	Pagination  Pagination `json:"pagination"`
}

type InvitationRecordResponse struct {
	ID               int64     `json:"id"`
	Code             string    `json:"code"`
	UsedBy           string    `json:"used_by"`
	UsedByExternalID string    `json:"used_by_external_id"`
	UsedAt           time.Time `json:"used_at"`
	Status           string    `json:"status"`
}

File: frontend/src/services/api/invitation.js
import {
  unaryCall,
  messages,
  invitationClient,
  baseToObject,
  handleResponse,
  mapInvitation,
  mapInvitationRecord
} from './shared';

const {
  ListInvitationsRequest,
  CreateInvitationRequest,
  UpdateInvitationRequest,
  DeleteInvitationRequest,
  ListInvitationRecordsRequest
} = messages;

export const listInvitations = async (params = {}) => {
  const req = new ListInvitationsRequest({
    maxUsedCnt: typeof params.max_used_cnt === 'number' ? params.max_used_cnt : undefined,
    expiresAtStart: params.expires_at_start || undefined,
    expiresAtEnd: params.expires_at_end || undefined,
    createdAtStart: params.created_at_start || undefined,
    createdAtEnd: params.created_at_end || undefined,
    disabled: typeof params.disabled === 'boolean' ? params.disabled : undefined,
    orderBy: params.order_by || undefined,
    orderDesc: typeof params.order_desc === 'boolean' ? params.order_desc : undefined,
    page: params.page || 1,
    pageSize: params.page_size || 20,
    onlyMine: typeof params.only_mine === 'boolean' ? params.only_mine : undefined,
    keyword: params.keyword || undefined
  });

  const resp = await unaryCall(invitationClient, 'listInvitations', req);
  const base = baseToObject(resp);
  return handleResponse(base, {
    invitations: (resp.invitations || []).map(mapInvitation).filter(Boolean),
    total: resp.total ?? 0
  });
};

export const createInvitation = async (params = {}) => {
  const req = new CreateInvitationRequest({
    maxUsedCnt: typeof params.max_used_cnt === 'number' ? params.max_used_cnt : undefined,
    expiresAt: params.expires_at || undefined,
    note: params.note || undefined
  });

  const resp = await unaryCall(invitationClient, 'createInvitation', req);
  const base = baseToObject(resp);
  return handleResponse(base, {
    invitation: resp.invitation ? mapInvitation(resp.invitation) : null
  });
};

export const updateInvitation = async (params = {}) => {
  const req = new UpdateInvitationRequest({
    code: params.code || '',
    maxUsedCnt: typeof params.max_used_cnt === 'number' ? params.max_used_cnt : undefined,
    expiresAt: params.expires_at || undefined,
    disabled: typeof params.disabled === 'boolean' ? params.disabled : undefined,
    note: params.note || undefined
  });

  const resp = await unaryCall(invitationClient, 'updateInvitation', req);
  const base = baseToObject(resp);
  return handleResponse(base, {});
};

export const deleteInvitation = async (code) => {
  const req = new DeleteInvitationRequest({ code: code || '' });

  const resp = await unaryCall(invitationClient, 'deleteInvitation', req);
  const base = baseToObject(resp);
  return handleResponse(base, {});
};

export const listInvitationRecords = async (params = {}) => {
  const req = new ListInvitationRecordsRequest({
    code: params.code || undefined,
    status: params.status || undefined,
    usedAtStart: params.used_at_start || undefined,
    usedAtEnd: params.used_at_end || undefined,
    page: params.page || 1,
    pageSize: params.page_size || 20,
    onlyMine: typeof params.only_mine === 'boolean' ? params.only_mine : undefined,
    keyword: params.keyword || undefined
  });

  const resp = await unaryCall(invitationClient, 'listInvitationRecords', req);
  const base = baseToObject(resp);
  return handleResponse(base, {
    records: (resp.records || []).map(mapInvitationRecord).filter(Boolean),
    total: resp.total ?? 0
  });
};
